### PR TITLE
Abstracted FFI to support multiple param-based VerkleTrie with same interface

### PR DESCRIPTION
**Pull request by realnimish**
The changes are passing the unit tests of the wrapper library. Need to check from the client side too!
Only change is in the API is `verkle_trie_new(byte, byte)` from `verkle_trie_new()`.

### CODE REVIEW

**Comment by realnimish** [comment]
If this works, then I will try to abstract the model code into a trait to avoid all the repetition in the codebase

File: src/lib.rs [hunk 1]
```diff
@@ -26,115 +22,95 @@ pub struct Proof {
...
-    let _db = MemoryDb::new();
-    let committer = TestCommitter;
-
-    let config = Config { db: _db, committer };
-    let mut _trie = Trie::new(config);
-    let mut vt = VerkleTrie {
-        trie: _trie,
+pub extern fn verkle_trie_new(database_scheme: u8, commit_scheme: u8) -> *mut VerkleTrie {
+    let vt = match database_scheme {
+        0 => match commit_scheme {
```
Comment (tanishqjasoria) [inline src/lib.rs]: add support for BareDiskMetalDb also

File: src/lib.rs [hunk 2]
```diff
@@ -298,7 +274,7 @@ mod tests {
 
     #[test]
     fn generate_proof_test(){
-        let trie = verkle_trie_new();
+        let trie = verkle_trie_new(0,0);
```
Comment (tanishqjasoria) [inline src/lib.rs]: add test cases for all the variants

File: src/verkle_variants/memory_prelagrange.rs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+            Ok(cnf) => cnf,
+            Err(_) => {
+                let _db = MemoryDb::new();
+                VerkleConfig::open(_db).unwrap()
+            },
+        };
+        let mut _trie = Trie::new(config);
+        _trie
+    }
+}
```
Comment (tanishqjasoria) [inline src/verkle_variants/memory_prelagrange.rs]: add a new line here

File: src/verkle_variants/memory_test.rs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+impl FFI for VerkleTrie {
+    fn verkle_trie_new() -> Self {
+        let _db = MemoryDb::new();
+        let committer = TestCommitter;
+        let config = Config{ db: _db, committer };
+        let mut _trie = Trie::new(config);
+        _trie
+    }
+}
```
Comment (tanishqjasoria) [inline src/verkle_variants/memory_test.rs]: newline

**Review by tanishqjasoria — CHANGES_REQUESTED** [review]
add the test cases and all the variants that were discussed

File: src/lib.rs (hunk 1 above)
Comment (realnimish) [reply src/lib.rs]: Do you mean RocksDB?

File: src/lib.rs (hunk 2 above)
Comment (realnimish) [reply src/lib.rs]: Which among these should I consider?

1. Should I add the unit tests in their respective files?
2. Should I have unit tests in the `lib.rs` covering all the combinations
3. Both

File: src/lib.rs (hunk 1 above)
Comment (tanishqjasoria) [reply src/lib.rs]: yes

File: src/lib.rs (hunk 2 above)
Comment (tanishqjasoria) [reply src/lib.rs]: adding in the individual files would be better I guess?

Comment (realnimish) [reply src/lib.rs]: I've added unit tests for each file but I'm also thinking of adding full coverage test in the `lib.rs` too. Maybe after that I can remove tests from individual files if they seem to do the same tests

**Comment by realnimish** [comment]
I believe all the parts have been implemented now. Waiting for the https://github.com/crate-crypto/rust-verkle/pull/47 to get merged before I can include the unit tests of PrecomputeLagrange commitment based VerkleTrie(s).

File: src/verkle_variants/rocksdb_prelagrange.rs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    database::VerkleDb,
+    committer::precompute::PrecomputeLagrange,
+};
+use crate::verkle_variants::traits::FFI;
+use rocksdb::{DB, WriteBatch};
+
+pub struct RocksDb(DB);
+pub struct _WriteBatch(WriteBatch);
+
+impl BareMetalDiskDb for RocksDb {
```
Comment (tanishqjasoria) [inline src/verkle_variants/rocksdb_prelagrange.rs]: what is the need of this, isnt this already done in the rust verkle repo?

Comment (tanishqjasoria) [reply src/verkle_variants/rocksdb_prelagrange.rs]: same for BareMetalDiskDb, BatchWriter, BatchDB, BareMetalKVDb?

**Review by tanishqjasoria — APPROVED** [review]
can be merged after the tests on the client side pass

File: src/verkle_variants/rocksdb_prelagrange.rs (hunk 5 above)
Comment (realnimish) [reply src/verkle_variants/rocksdb_prelagrange.rs]: I tried to use the one defined in their repo but it didn't worked. I believe it's there for reference purpose.

Comment (tanishqjasoria) [reply src/verkle_variants/rocksdb_prelagrange.rs]: https://github.com/realnimish/rust-verkle-wrapper/pull/6

**Review by tanishqjasoria — APPROVED** [review]

**Pull request merged**